### PR TITLE
new protocol: transfer xid when use java thread

### PR DESCRIPTION
**Pull request by Justice-love**
### Ⅰ. Describe what this PR did
transfer xid when use java thread.
if  i new a thread or use thread pool to run a thread in global transaction, xid will not transfer into the new thread and  i want the xid transfer into the new thread transparent.

* How do I implement it?
1. use aspectj to extends the subclass of Runnable and Callbale, add a field to store xid.
2. extends constructor of Runnable and Callbale subclass,  get  xid in the current thread , store use the xid field
3. extends run() and call(), before the logic, find  value of xid field  and call RootContext.bind(xid)

### CODE REVIEW

**Comment by Justice-love** [comment]
refe [aspectj-maven-plugin pr #45](https://github.com/mojohaus/aspectj-maven-plugin/pull/45)

**Comment by Justice-love** [comment]
> add this dependency in seata-all

done

File: integration/java-thread/pom.xml [hunk 1]
```diff
@@ -0,0 +1,104 @@
...
+
+    <build>
+        <plugins>
+            <plugin>
+                <groupId>com.nickwongdev</groupId>
+                <artifactId>aspectj-maven-plugin</artifactId>
+                <version>1.12.1</version>
+                <configuration>
+                    <complianceLevel>1.8</complianceLevel>
+                    <source>1.8</source>
```
Comment (slievrly) [inline integration/java-thread/pom.xml]: ${maven.compiler.source}

File: integration/java-thread/pom.xml [hunk 2]
```diff
@@ -0,0 +1,104 @@
...
+                    <target>1.8</target>
```
Comment (slievrly) [inline integration/java-thread/pom.xml]: ${maven.compiler.target}

File: integration/java-thread/pom.xml [hunk 3]
```diff
@@ -0,0 +1,104 @@
...
+                    <showWeaveInfo>true</showWeaveInfo>
+                    <verbose>true</verbose>
+                    <Xlint>ignore</Xlint>
+                    <encoding>UTF-8</encoding>
```
Comment (slievrly) [inline integration/java-thread/pom.xml]: ${project.build.sourceEncoding}

File: all/pom.xml [hunk 4]
```diff
@@ -182,6 +182,11 @@
...
             <version>${project.version}</version>
         </dependency>
+        <dependency>
+            <groupId>io.seata</groupId>
+            <artifactId>seata-java-thread</artifactId>
+            <version>${project.version}</version>
+        </dependency>
         <dependency>
             <groupId>io.seata</groupId>
             <artifactId>seata-rm</artifactId>
```
Comment (slievrly) [inline all/pom.xml]: should add aspectjrt、aspectjweaver in all.pom? bom just as dependencyManagement

File: integration/java-thread/src/test/java/io/seata/integration/thread/ContextTest.java [hunk 5]
```diff
@@ -0,0 +1,75 @@
...
+    }
+
+    @Test
+    public void nonPropagateRunnable() throws InterruptedException {
+
+        RootContext.bind("none-context");
+        executorService.submit(new NonPropagateRunnable() {
+            @Override
+            public void run() {
+                Assertions.assertEquals(RootContext.getXID(), "none-context");
```
Comment (slievrly) [inline integration/java-thread/src/test/java/io/seata/integration/thread/ContextTest.java]: should not equal?

File: integration/java-thread/src/test/java/io/seata/integration/thread/ContextTest.java [hunk 6]
```diff
@@ -0,0 +1,75 @@
...
+            }
+        });
+        executorService.submit(new NonPropagateCallable<Object>() {
+            @Override
+            public Object call() throws Exception {
+                Assertions.assertEquals(RootContext.getXID(), "none-context");
+                return null;
```
Comment (slievrly) [inline integration/java-thread/src/test/java/io/seata/integration/thread/ContextTest.java]: should not equal?

**Comment by Justice-love** [comment]
resolved problem and add a config for this feature

**Comment by Justice-love** [comment]
> Say my understanding:
> 
> 1. Create two entry points through aspectj, namely init() and cut(), init() is used to save the main thread's xid to CallableTransactionPropagation#xid after thread initialization, and cut() is used in the calling thread Before the call method, put the xid saved in CallableTransactionPropagation#xid into the current thread threadLocal for subsequent getXID() acquisition
> 2. If you don't need thread transfer, you need to implement NonPropagateCallable or NonPropagateRunnable
> 3. There is a doubt that the threadLocal cache of these threads seems to have no place to clear
> 4. CallableTransactionPropagationAspect and RunnableTransactionPropagationAspect have a lot of redundant code, can they be combined to deal with it?

1. 首先需要区分线程和runnable, 线程是可重复利用的, 比如在线程池场景, 但runnable的实例往往是一次性的
2. 需要了解runnable的不通代码块在对应哪个线程中执行, 构造函数在父线程中执行, run()方法在子线程中执行, 而该方案是利用runnable实例的不同代码块在不同线程中执行的特性来传递xid

所以, 你提的问题关于threadLocal的缓存问题, 由于并不涉及线程, runnable实例也是一次性的, 所以应该不需要清理

**Comment by a364176773** [comment]
resolve code conflicts

**Comment by funky-eyes** [comment]
Due to code conflicts and age, if you are still interested in participating in the community, please resubmit this pr to the 2.x branch after resolving the conflict.
由于代码冲突以及年代久远，如果你还有兴趣参与到社区，请解决完冲突后重新请提交这份pr至2.x分支中

**Pull request closed without merging**